x-pack/filebeat/input/httpjson/request.go

```diff
@@ -341,11 +378,22 @@ func (r *requester) doRequest(stdCtx context.Context, trCtx *transformContext, p
 				}
 				resps = intermediateResps
 			}
-			n += r.processAndPublishEvents(stdCtx, trCtx, publisher, resps, i < len(r.requestFactories), i)
+
+			var events <-chan maybeMsg
+			if rf.isChain {
+				events = rf.chainResponseProcessor.startProcessing(stdCtx, trCtx, resps)
+			} else {
+				events = r.responseProcessors[i].startProcessing(stdCtx, trCtx, resps)
+			}
+			n += processAndPublishEvents(trCtx, events, publisher, i < len(r.requestFactories), r.log)
 		}
 	}
 
 	defer httpResp.Body.Close()
```

Comment:
Do you know why this is here? I think it is a relic from before the step processing was added and there was only a single response to deal with. Do you think it is reasonable to remove it? (Probably not in this change).

Reply:
@efd6 Unfortunately there are no tests for the chain requests as a whole , whenever it was written ,  it was done so without tests for the actual request cycle. They will need to be introduced from scratch. As for the **defer httpResp.Body.Close()** , it is definitely an artifact of the past, but sadly the last instance of the response body still remains open so we definitely need to defer this / or close gracefully , hence we cannot remove this atm without changing how the code works.

Reply:
**Added tests for this scenario**

Reply:
> the last instance of the response body still remains open so we definitely need to defer this / or close gracefully , hence we cannot remove this atm without changing how the code works.

I don't think this is the case; the body is always a product of `*httpClient.do` which always dumps the body into a `*bytes.Buffer` wrapped in an `io.nopCloser` on success, or closing in the case of a failure during the copy (or it returns an `http.NoBody` if empty). So all the closes of the body after `*httpClient.do` returns are futile. The reason to close a resp is to free the netconn, but that has already happened.

Reply:
yea , my bad .. just saw the client.do & the drain body methods in detail and the connections are always released. Then it doesn't matter if **httpResp.Body.Close()** is there or not, we could just remove it.